apollo-runtime/src/appleMain/kotlin/com/apollographql/apollo3/network/http/StreamingNSURLSessionHttpEngine.kt

```diff
@@ -0,0 +1,313 @@
+            setValue(requestBody.contentLength.toString(), forHTTPHeaderField = "Content-Length")
+          }
+          val body = Buffer().apply { requestBody.writeTo(this) }.readByteArray().toNSData()
+          setHTTPBody(body)
+        }
+      }
+      setCachePolicy(NSURLRequestReloadIgnoringCacheData)
+    }
+
+    val httpDataPipe = Pipe()
+    val httpDataSink = httpDataPipe.sink.buffer()
+    val httpDataSource = httpDataPipe.source.buffer()
+
+    val handler = object : StreamingDataDelegate.Handler {
+      override fun onResponse(response: NSHTTPURLResponse) {
+        mainContinuation.resumeWith(
+            buildHttpResponse(
+                data = httpDataSource,
+                httpResponse = response,
+                error = null,
+            )
+        )
+      }
+
+      override fun onData(data: NSData) {
+        httpDataSink.write(data.toByteString())
+        httpDataSink.flush()
+      }
+
+      override fun onComplete(error: NSError?) {
+        httpDataSink.close()
+        if (error != null) mainContinuation.resumeWith(
+            buildHttpResponse(
+                data = httpDataSource,
+                httpResponse = null,
+                error = error,
+            )
+        )
+      }
+    }
+
+    val task = nsUrlSession.dataTaskWithRequest(nsMutableURLRequest)
+    delegate.registerHandlerForTask(task, handler)
+
+    invokeOnCancellation {
+      task.cancel()
+    }
+    task.resume()
+  }
+
+  override fun dispose() {
+  }
+}
+
+private fun buildHttpResponse(
+    data: BufferedSource,
+    httpResponse: NSHTTPURLResponse?,
+    error: NSError?,
+): Result<HttpResponse> {
+
+  if (error != null) {
+    return Result.failure(
+        ApolloNetworkException(
+            message = "Failed to execute GraphQL http network request",
+            platformCause = error
+        )
+    )
+  }
+
+  if (httpResponse == null) {
+    return Result.failure(
+        ApolloNetworkException("Failed to parse GraphQL http network response: EOF")
+    )
+  }
+
+  val httpHeaders = httpResponse.allHeaderFields
+      .map { (key, value) ->
+        HttpHeader(key.toString(), value.toString())
+      }
+
+  val statusCode = httpResponse.statusCode.toInt()
+
+  return Result.success(
+      HttpResponse.Builder(statusCode = statusCode)
+          .headers(httpHeaders)
+          .body(bodySource = data)
+          .build()
+  )
+}
+
+/**
+ * Order of the callbacks in the no-error case:
+ * - didReceiveResponse
+ * - didReceiveData
+ * - didReceiveData
+ * ...
+ * - willCacheResponse
+ *
+ * Error case (e.g. no connectivity):
+ * - didCompleteWithError (with a non-null error)
+ */
+private class StreamingDataDelegate : NSObject(), NSURLSessionDataDelegateProtocol {
+  interface Handler {
+    fun onResponse(response: NSHTTPURLResponse)
+    fun onData(data: NSData)
+    fun onComplete(error: NSError?)
+  }
+
+  private val handlers = mutableMapOf<NSURLSessionTask, Handler>()
```

Comment:
I'm guessing you could create a new nsUrlSession and a new NSURLSessionDataDelegateProtocol for every execute() call and that'd save this indirection and make things a bit more immutable ?

Reply:
That would simplify the code but I think that we're not supposed to create a new session per request, for performance - a bit like we wouldn't create a new OkHttpClient every time. For instance I guess it can keep/re-use certain connections alive.

[Here](https://developer.apple.com/forums/thread/84663) I found an answer recommending against it.